xdsl/frontend/passes/desymref.py

```diff
@@ -0,0 +1,389 @@
+                        self.rewriter.replace_op(read, [], [write.operands[0]])
+
+    def _prune_unused_reads(self, block: Block):
+        is_unused_read: Callable[
+            [Operation],
+            bool] = lambda op: is_read(op) and len(op.results[0].uses) == 0
+        unused_reads = select_ops_by(block, is_unused_read)
+        for read in unused_reads:
+            self.rewriter.erase_op(read)
+
+    def prune_uses_without_definitions(self, block: Block):
+        """Removes all possible symbol uses in a single block."""
+        prepared_symbols: Set[str] = set()
+
+        while True:
+            self._prune_unused_reads(block)
+
+            # Find all symbols that are still in use in this block.
+            symbol_worklist: Set[str] = set(
+                filter(lambda symbol: symbol not in prepared_symbols,
+                       get_symbols(block)))
+            if len(symbol_worklist) == 0:
+                return
+
+            for symbol in symbol_worklist:
+                reads = select_ops_by(
+                    block, lambda op: is_read(op) and get_symbol(op) == symbol)
+                writes = select_ops_by(
+                    block,
+                    lambda op: is_write(op) and get_symbol(op) == symbol)
+                assert len(reads) > 0 or len(writes) > 0
+
+                # There are no reads, so we can only keep the last write to the
+                # symbol.
+                if len(reads) == 0:
+                    for write in writes[:-1]:
+                        self.rewriter.erase_op(write)
+                    prepared_symbols.add(symbol)
+                    continue
+
+                # There are no writes, so we can replace all reads with this
+                # symbol.
+                if len(writes) == 0:
+                    for read in reads[1:]:
+                        self.rewriter.replace_op(read, [],
+                                                 [reads[0].results[0]])
+                    prepared_symbols.add(symbol)
+                    continue
+
+                # Sets of reads and writes are disjoint.
+                last_read_idx = block.get_operation_index(reads[-1])
+                first_write_idx = block.get_operation_index(writes[0])
+                if last_read_idx < first_write_idx:
+                    for read in reads[1:]:
+                        self.rewriter.replace_op(read, [],
+                                                 [reads[0].results[0]])
+                    for write in writes[:-1]:
+                        self.rewriter.erase_op(write)
+                    prepared_symbols.add(symbol)
+                    continue
+
+                # Otherwise, replace reads with the closest preceding write.
+                for read in reads:
+                    write = lower_bound(writes, read)
+                    if write is not None:
+                        self.rewriter.replace_op(read, [], [write.operands[0]])
+
+
+@dataclass
+class DesymrefyPass:
```

Comment:
Atm, this could just e a function, right? I guess there is a future reason that you have it in a class?

Reply:
Sort of. In the future, for multiple-block CFGs we need to compute dominance info, which would be done inside this `run` function.

The reason for it being a class is if we want to define a parent `Pass` class it is easier to adapt. Then a common interface would be `run(op: Operation)`, very similar to what LLVM does with its passes.